### PR TITLE
Update `models.transformers` to use `SequenceGeneratorAdapter` and `OutlinesLogitsProcessors`

### LINKED ISSUE #789: RegexPrefixAllowedTokens does not work for batch
### Describe the issue as clearly as possible:

In these lines of code, https://github.com/outlines-dev/outlines/blob/3a41b0edf389ceba49126e811ed0c1b23b50f235/outlines/integrations/transformers.py#L87-L100

it seems that the batch id is not used at all, and there is an error triggered when batch is used and input_ids is empty.
https://github.com/outlines-dev/outlines/blob/3a41b0edf389ceba49126e811ed0c1b23b50f235/outlines/integrations/transformers.py#L116

Do you know how this can be fixed? Thanks! @saattrupdan @rlouf

### Steps/code to reproduce the bug:

```python
# Based on https://github.com/outlines-dev/outlines/blob/3a41b0edf389ceba49126e811ed0c1b23b50f235/examples/transformers_integration.py#L6
from pydantic import BaseModel
from transformers import  AutoModelForCausalLM, AutoProcessor

from outlines.integrations.transformers import JSONPrefixAllowedTokens


class Person(BaseModel):
    first_name: str
    surname: str


processor = AutoProcessor.from_pretrained("mistralai/Mistral-7B-v0.1")
model = AutoModelForCausalLM.from_pretrained("mistralai/Mistral-7B-v0.1")
prefix_allowed_tokens_fn = JSONPrefixAllowedTokens(
    schema=Person, tokenizer_or_pipe=processor.tokenizer,
)

inputs = processor(...)
model.generate(
    **inputs,
    prefix_allowed_tokens_fn=prefix_allowed_tokens_fn,
)
```

### Expected result:

```shell
When inputs is a batch, this should not trigger an empty list indexing error
```

### Error message:

_No response_

### Outlines/Python version information:

Version information
<details>
```
(command output here)
```
</details>

### Context for the issue:

_No response_

### LINKED ISSUE #806: Update the `transformers` integration
In the vein of #782 and #772 we should refactor the `transformers` integration to use logits processor.

We will keep the custom sampling loop, but via a `Outlines` model that wraps `transformers` models. We should be able to remove `torch` and `transformers` as default dependencies as a result.

### PR DESCRIPTION
**Pull request by lapp0**
~~In draft until https://github.com/huggingface/transformers/pull/31448 makes it into a new `transformers` release~~

Fixes https://github.com/outlines-dev/outlines/issues/1021

Fixes #789

Fixes #806 (does everything except remove the issues requirement of "remove `torch`")

Closes https://github.com/outlines-dev/outlines/pull/910 (device inconsistency issue handled through other means)

# Problem

- For #789 in `SequenceGenerator` if `input_ids` is empty, it fails.
- Generally, `models.transformers` has an inconsistent implementation with other modules making the codebase harder to manage.

# Solution

- Use `SequenceGeneratorAdapter` for transformers instead of `SequenceGenerator`
- Implement `Transformers.generate` and `Transformers.stream` which use `model.generate(logits_processor=...)` argument with `outlines.processors.OutlinesLogitsProcessor`

### Additional Changes

- ~~Temporarily disables `stop_at` argument for transformers and implements test to determine whether upstream is fixed (https://github.com/huggingface/transformers/issues/31435)~~
  - fixed upstream, change reverted
- Update docs and examples to use `seed` instead of `rng`
- Disables `outlines.generate.cfg` for now (https://github.com/outlines-dev/outlines/issues/959)

# TODO:

- [x] Determine cause of regression when generating json
  - was caused by `GenerationConfig` default `max_tokens` being 20
- [x] replace `rng` with `seed` in all tests and all documentation.
- [x] Awaiting `transformers` fix for https://github.com/huggingface/transformers/issues/31435 so `stop_at` works.

### CODE REVIEW

File: outlines/generate/cfg.py [hunk 1]
```diff
@@ -24,40 +22,16 @@ def cfg(model, cfg_str: str, sampler: Sampler = multinomial()) -> SequenceGenera
...
-
-
-@cfg.register(MLXLM)
-@cfg.register(VLLM)
-def cfg_unimplemented(
-    model,
-    cfg_str: str,
-    sampler: Sampler = multinomial(),
-):
     raise NotImplementedError(
```
Comment (lapp0) [inline outlines/generate/cfg.py:28]: Not directly related to the PR, but because `cfg` is broken for all models, we should reference the issue to enable it.

File: outlines/generate/regex.py [hunk 2]
```diff
@@ -39,8 +40,9 @@ def regex(model, regex_str: str, sampler: Sampler = multinomial()):
 
 
 @regex.register(MLXLM)
-def regex_mlxlm(
-    model: MLXLM,
+@regex.register(Transformers)
```
Comment (lapp0) [inline outlines/generate/regex.py:43]: the `_unified` dispatchers will become the default dispatcher as a next step. In this PR it's just used by `MLXLM` and `Transformers`

File: outlines/models/transformers.py [hunk 3]
```diff
@@ -190,6 +194,169 @@ def __call__(
...
+        from transformers import GenerationConfig, set_seed
+
+        max_tokens, stop_at, seed = dataclasses.astuple(generation_parameters)
+        sampler, num_samples, top_p, top_k, temperature = dataclasses.astuple(
+            sampling_parameters
+        )
+
+        # global seed, not desirable
+        if seed is not None:
+            set_seed(seed)
```
Comment (lapp0) [inline outlines/models/transformers.py:214]: We might consider getting rid of `seed` all-together in a separate issue.

File: tests/generate/conftest.py [hunk 4]
```diff
@@ -1,24 +1,40 @@
...
+        skipped_models.append("model_vllm")
+
+    skip_marker = pytest.mark.skip(
+        reason="Skipping test due to incompatible hardware / system."
+    )
+    for item in items:
+        if "model_fixture" in item.fixturenames:
+            model_param = item.callspec.params.get("model_fixture", None)
+            if model_param in skipped_models:
+                item.add_marker(skip_marker)
```
Comment (lapp0) [inline tests/generate/conftest.py:37]: These changes aren't directly related to the PR, but a minor change: I added vLLM integration tests to `test_generate.py`. This disables vLLM (and mlxlm) tests if the hardware is wrong.

File: examples/llamacpp_example.py [hunk 5]
```diff
@@ -37,10 +36,9 @@ class Character(BaseModel):
     generator = outlines.generate.json(model, Character)
 
     # Draw a sample
-    rng = torch.Generator(device="cpu")
-    rng.manual_seed(789005)
+    seed = 789005
 
     prompt = "Instruct: You are a leading role play gamer. You have seen thousands of different characters and their attributes.\nPlease return a JSON object with common attributes of an RPG character. Give me a character description\nOutput:"
 
-    sequence = generator(prompt, rng=rng, max_tokens=512)
```
Comment (lapp0) [inline examples/llamacpp_example.py:45]: This documentation change reflects behavior in `main`.

File: docs/reference/models/transformers.md [hunk 6]
```diff
@@ -30,4 +30,55 @@ tokenizer = AutoTokenizer.from_pretrained("gpt2")
 model = models.Transformers(llm, tokenizer)
 ```
 
+# Using Logits Processors
```
Comment (rlouf) [inline docs/reference/models/transformers.md:33]: We'll need to improve the documentation to reach something similar to the lamacpp integration's.

File: outlines/models/transformers.py [hunk 7]
```diff
@@ -179,6 +182,10 @@ def forward(
 
         return output.logits, output.past_key_values
 
+    @property
+    def device(self):
+        return self.model.device
```
Comment (lapp0) [inline outlines/models/transformers.py:187]: test should be fixed rather than adding this unnecessary property

File: outlines/models/transformers.py [hunk 8]
```diff
@@ -190,6 +197,182 @@ def __call__(
...
+    def stream(
+        self,
+        prompts: Union[str, List[str]],
+        generation_parameters: GenerationParameters,
+        logits_processor: Optional["OutlinesLogitsProcessor"],
+        sampling_parameters: SamplingParameters,
+    ) -> Iterator[Union[str, List[str]]]:
+        """
+        Temporary stream stand-in which implements stream() signature
+        and equivalent behaviour but isn't iterable.
```
Comment (lapp0) [inline outlines/models/transformers.py:261]: Incorrect doc string

File: docs/reference/models/transformers.md (hunk 6 above)
Comment (lapp0) [reply docs/reference/models/transformers.md:33]: We should plan a restructuring and cleaning up of documentation in a separate issue. I could share some ideas in call on how we might approach this.

In this case, a lot of information documented for `llamacpp` applies to all other models including `transformers`. We shouldn't repeat ourselves. We should explain the behavior of all models generally, highlight the models differences with a feature table, and document only `transformers` specific information on its documentation page.

Comment (rlouf) [reply docs/reference/models/transformers.md:33]: What I meant was listing the main arguments to you can pass when initialising and calling the model, cf https://outlines-dev.github.io/outlines/reference/models/llamacpp/

**Review by rlouf — APPROVED** [review]

**Comment by rlouf** [comment]
Very good work, thank you!
